crates/items/src/item_stack.rs

```diff
@@ -0,0 +1,171 @@
+#![forbid(unsafe_code, warnings)]
+
+use crate::Enchantment;
+
+// TODO Remove when all codegen has been migrated and generated. DUMMY STRUCT
+#[derive(Copy, Clone)]
+pub struct Item {}
+impl Item {
+    pub fn durability(&self) -> Option<u32> { None }
+    pub fn stack_size(&self) -> u32         { 64 }
+}
+
+/// Represents an item stack.
+///
+/// An item stack includes an item type, an amount and a bunch o properties (enchantments, etc.)
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ItemStack {
+
+    // TODO Migrate all the codegen boilerplate.
+    /// The item type of this `ItemStack`.
+    pub item: Item,
+
+    /// The number of items in the `ItemStack`.
+    pub count: u32,
+
+    /// Damage to the item, if it's damageable.
+    pub damage: Option<u32>,
+
+    /// Repair cost of the item, if it's repairable.
+    pub repair_cost: Option<u32>,
+
+    /// Enchantments applied to the item.
+    pub enchantments: Vec<Enchantment>,
+
+    // TODO Add other properties like title and lore.
+}
+
+impl ItemStack {
+
+    /// Creates a new `ItemStack`
+    pub fn new(item: Item, count: u32) -> Self {
+        Self {
+            item,
+            count,
+            damage: item.durability().map(|_| 0)
+        }
+    }
+
+    /// Returns whether the given item stack has
+    /// the same type as (but not necessarily the same
+    /// amount as) `self`.
+    pub fn has_same_type(&self, other: &Self) -> bool {
+        self.item == other.item
+    }
+
+    /// Returns whether the given item stack has the same damage
+    /// as `self`.
+    pub fn has_same_damage(&self, other: &Self) -> bool {
+        self.damage == other.damage
+    }
+
+    /// Returns whether the given `ItemStack` has
+    /// the same count as (but not necessarily the same
+    /// type as) `self`.
+    pub fn has_same_count(&self, other: &Self) -> bool {
+        self.count == self.count
+    }
+
+    /// Returns whether the given `ItemStack` has the same
+    /// type and count as `self`.
+    pub fn has_same_type_and_count(&self, other: &Self) -> bool {
+        self.item == other.item && self.count == other.count
+    }
+
+    /// Returns whether the given `ItemStack` has
+    /// the same type and damage as `self`.
+    pub fn has_same_type_and_damage(&self, other: &Self) -> bool {
+        self.item == other.item && self.damage == other.damage
+    }
+
+    /// Returns the item type for this `ItemStack`.
+    pub fn item(&self) -> Item {
+        self.item
+    }
+
+    /// Returns the number of items in this `ItemStack`.
+    pub fn count(&self) -> u32 {
+        self.count
+    }
+
+    /// Adds more items to this `ItemStack`. Returns the new count.
+    pub fn add(&mut self, count: u32) -> u32 {
+        self.count += count;

```

Comment:
Stacks should be limited.

Reply:
We have two possible approaches:
* Use `Result<u32, ()>` and fail when trying to add items to a full-stack.
* Ignore the request and return the current count.
I personally think the first one is better.

Reply:
Think we should use an option instead of `Result`